lib/logstash/inputs/file.rb

```diff
@@ -210,126 +261,127 @@ def register
       end
 
       #pick SINCEDB_DIR if available, otherwise use HOME
-      sincedb_dir = ENV["SINCEDB_DIR"] || ENV["HOME"]
+      sincedb_dir = Pathname.new(ENV["SINCEDB_DIR"] || ENV["HOME"])
 
       # Join by ',' to make it easy for folks to know their own sincedb
       # generated path (vs, say, inspecting the @path array)
-      @sincedb_path = File.join(sincedb_dir, ".sincedb_" + Digest::MD5.hexdigest(@path.join(",")))
+      @sincedb_path = sincedb_dir.join(".sincedb_" + Digest::MD5.hexdigest(@path.join(",")))
 
       # Migrate any old .sincedb to the new file (this is for version <=1.1.1 compatibility)
-      old_sincedb = File.join(sincedb_dir, ".sincedb")
-      if File.exists?(old_sincedb)
-        @logger.debug("Renaming old ~/.sincedb to new one", :old => old_sincedb,
-                     :new => @sincedb_path)
-        File.rename(old_sincedb, @sincedb_path)
+      old_sincedb = sincedb_dir.join(".sincedb")
+      if old_sincedb.exists?
+        @logger.debug("Renaming old ~/.sincedb to new one", :old => old_sincedb.to_s,
+                     :new => @sincedb_path.to_s)
+        old_sincedb.rename(@sincedb_path)
       end
 
       @logger.info("No sincedb_path set, generating one based on the file path",
-                   :sincedb_path => @sincedb_path, :path => @path)
+                   :sincedb_path => @sincedb_path.to_s, :path => @path)
     end
 
-    if File.directory?(@sincedb_path)
+    if @sincedb_path.directory?
       raise ArgumentError.new("The \"sincedb_path\" argument must point to a file, received a directory: \"#{@sincedb_path}\"")
     end
 
-    @tail_config[:sincedb_path] = @sincedb_path
-
-    if @start_position == "beginning"
-      @tail_config[:start_new_files_at] = :beginning
+    @filewatch_config[:sincedb_path] = @sincedb_path
+
+    @filewatch_config[:start_new_files_at] = @start_position.to_sym
+
+    if @file_completed_action == 'log'
+      if @file_completed_log_path.nil?
+        raise ArgumentError.new('The "file_completed_log_path" setting must be provided when the "file_completed_action" is set to "log"')
+      else
+        @file_completed_log_path = Pathname.new(@file_completed_log_path)
+        unless @file_completed_log_path.exist?
+          begin
+            FileUtils.touch(@file_completed_log_path)
+          rescue
+            raise ArgumentError.new("The \"file_completed_log_path\" file can't be created: #{@file_completed_log_path}")
+          end
+        end
+      end
     end
 
+    if tail_mode?
+      require "filewatch/observing_tail"
```

Comment:
do we need dynamic class loading here? I usually try to avoid that unless there is a good reason to. sometime, depending on test coverage it will hide errors in ont or the other file.